iree/turbine/kernel/wave/utils.py

```diff
@@ -1043,3 +1044,208 @@ def evaluate_with_assumptions(constraints: list[Constraint], expr: IndexExpr) ->
     if isinstance(result, sympy.logic.boolalg.BooleanAtom):
         return False
     return True if any([result.equals(x) for x in facts]) else None
+
+
+def _get_start_index(i: IndexSequence | IndexExpr) -> IndexExpr:
+    if isinstance(i, IndexSequence):
+        i = i.start
+
+    return i
+
+
+def _get_start_indices(
+    src_indices: dict[IndexExpr, IndexSequence | IndexExpr]
+) -> list[IndexExpr]:
+    start_indices = []
+    for dim_indexing in src_indices:
+        i = _get_start_index(src_indices[dim_indexing])
+        start_indices.append(i)
+
+    return start_indices
+
+
+def _simplify_sympy_expr(expr: IndexExpr) -> IndexExpr:
```

Comment:
<img width="759" alt="Screenshot 2024-12-03 at 3 16 46 PM" src="https://github.com/user-attachments/assets/ed931e54-acbe-420b-8ba7-0da44aec63d2">
 How about something like this?

Reply:
To accomplish this, you would do the following
```
> next_indices[2] - prev_indices[2]
-floor(floor((Mod($T0, 4))/80 + floor($ARGC*HF*WF*BLOCK_K/8)/80)/3) + floor(floor((Mod($T0, 4))/80 + floor($ARGC*HF*WF*BLOCK_K/8)/80 + 1/640)/3)

sympy.cse(next_indices[2] - prev_indices[2])[0]
[(x0, (Mod($T0, 4))/80 + floor($ARGC*HF*WF*BLOCK_K/8)/80)]
sympy.cse(next_indices[2] - prev_indices[2])[1]
[-floor(floor(x0)/3) + floor(floor(x0 + 1/640)/3)]
```

Reply:
So, as any sane person should do, I wrote a fuzzing tests (see latest commit) and it indeed found some failing cases. But after stricten some checks (and marking induction vars as non-negative) and running 100 000s tests I'm now reasonable sure my transformations are correct.

Reply:
Thanks that sounds good, but I think we have been thinking about this the wrong way.

So essentially what we are looking to do here is determine whether the mapping is "contiguity-preserving" (which means that elements that are contiguous pre-mapping will remain contiguous post-mapping). This is a function entirely of the mapping and should not depend at all on the particular access pattern of the underlying data. So I think these patterns to simplify the index sequences are not required and instead we should be looking just at the mapping. For example, looking at one IGEMM mapping which maps a tensor of shape [M, K] to a tensor of shape [N, H, W, C]

```
x_mapping = tkw.IndexMapping(
        num_iterators=2,
        inputs={
            N: i // SZ_OUT,
            C: j % C,
            H: (i % SZ_OUT) % W_OUT * stride + (j // C) % WF,
            W: (i % SZ_OUT) // W_OUT * stride + (j // C) // WF,
        },
        outputs={M: i, K: j},
    )
```

we can immediately deduce that this mapping is contiguity-preserving because j gets mapped to j % C.

Reply:
yeah, `sympy.cse` looks useful, but I weren't able construct working code quickly. I will need to think more on it offline.

Reply:
sounds good, let's land this PR. I have a WIP PR that should simplify this a lot use sympy.cse.